### ISSUE TITLE
Question about ThrowOnUnexpectedArgs = false and argument ordering

### ISSUE THREAD

**Issue by ossunatec** [natemcmaster/CommandLineUtils#171]
Hi!

I know that using "ThrowOnUnexpectedArgs = false" makes the library stop parsing when an unknown arg is found and all the remaining arguments are copied to "RemainingArguments".

Is there any option to try parsing all arguments and then obtaining those that weren't valid? (not stopping at first invalid). If there's no such option would you consider adding it? I'd find it pretty useful when some of the arguments are to be passed to another tool and not to be consumed by my app.

Thanks!

**Comment by TheOneTrueAnt** [comment]
I hit a similar issue and in my case I found creating an `Argument` at position 0 of type `string[]` basically collected all of the unexpected arguments from the commandline into it. In my case I wanted to supply a variable number of arguments potentially before the options so this worked well for me. This means that a commandline like:
```
myapp foo bar --opt val baz
```
Will result in my `Opt` property have the value `val` and my property marked as `Argument(0)` having the value `{"foo","bar","baz"}`.  (I'm using Attribute-based parsing)

Note this seems to take precedence over `SingleOrNoValue` type options; if a value is given they will end up in the arguments "bucket" rather than assigned to the option.

Hopefully that's of some help in your situation.

**Comment by natemcmaster** [comment]
> Is there any option to try parsing all arguments and then obtaining those that weren't valid? If there's no such option would you consider adding it? 

Yes. In 3.0, I was considering replacing ThrowOnUnexpectedArgs with an enum. I skteched out an API for something like this a while ago (names are just a proposal), but I'd like to finish up 2.3 work before I tackle 3.0 and breaking changes like this.

```c#
public enum UnrecognizedArgumentHandling
{
   Throw, // the default
   StopProcessingAndCollectRemainingArgs, // equivalent to ThrowOnUnexpectedArgs = false
   ContinueAndCollectUnrecognizedArgs, // the behavior you're asking for.
}

public class ParseResult
{
   public string[] UnrecognizedArguments { get; }
}
```

In the meantime, @TheOneTrueAnt's suggestion is a really great way to work around the current behavior.

**Comment by SteveBenz** [comment]
I've got a similar need to the one described above, I thought I'd describe it here rather than open a new issue because it seems like the same code could solve both.  What I've got is something very much like "cmd /c …".  In that case, everything after the "/c" is going to be eaten up as a string array.  Now, I know what you're thinking, "Hey bud, I've already got you covered!  [Option]string[] C!"  Ah, but wait...  cmd also supports a "/q" parameter "cmd /c otherprogram /q" is not the same as "cmd /q /c otherprogram".

The other thing I'd like to add is that I think if you have to resort to UnrecognizedArguments, you're not in a good place.  It seems to me like ideally we'd find a way to capture these oddball cases through some kind of argument-parsing extension...  Some way where we can say "look, when you see this option or argument, let this function have a look at the remainder of the argument list and eat up what it can handle."

**Comment by natemcmaster** [comment]
Stale bot closed this, but I'd like to revisit this in 3.0. Reopening

**Comment by natemcmaster** [comment]
Finally got around to this. Sorry it took so long, but it required making some breaking changes. Here's a proposal to implement this: https://github.com/natemcmaster/CommandLineUtils/pull/348

**Comment by natemcmaster** [comment]
Resolved in #348. Released in 3.0 alpha.
